lib/src/commit_builder.rs

```diff
@@ -21,8 +21,9 @@ use crate::repo::MutableRepo;
 use crate::settings::UserSettings;
 
 #[must_use]
-pub struct CommitBuilder<'repo> {
+pub struct CommitBuilder<'repo, 'settings> {
     mut_repo: &'repo mut MutableRepo,
+    settings: &'settings UserSettings,
```

Comment:
It's probably simpler to pass in the `&UserSettings` to `generate_new_change_id()` or what do you think? It's a bit redundant for the callers since they already passed passed it in when they created the `CommitBuilder`, but it keeps this struct simpler.

Reply:
That's what I did originally, and I didn't like the redundancy. One way to make the struct slightly simple would be to use the same lifetime for `mut_repo` and `settings`. I'm happy with any of these three options.

Reply:
I don't like to restrict lifetimes unnecessarily, so I don't like that option. I actually was going to suggest relaxing that restriction even in the tests where you had `UserSettings<'repo, 'repo>` because I found it confusing.

Since there are so few callers of `generate_new_change_id()`, I don't mind the redundancy much, but I'm also okay with the current form.

Reply:
> . I actually was going to suggest relaxing that restriction even in the tests where you had UserSettings<'repo, 'repo> because I found it confusing.

The compiler forced those on me. That made me think that the extra flexibility from the different lifetimes is illusory.

I'll just pass UserSettings to `generate_new_change_id` again, I think.

Reply:
Since rng is behind `Arc<Mutex>`, maybe it (or its wrapper like ChangeIdGenerator) can be just cloned to `CommitBuilder`?

Reply:
True, that's probably the easiest way of doing it. I think it's only behind an `Arc` because we currently clone `UserSettings` in a few places. I'd like to avoid doing that because it makes it confusing what the source of truth is, but it may still be useful to have the rng be behind an `Arc`.

Reply:
The RNG currently needs access to the settings on the first call of the `gen` function. That seemed worth it so that all RNGs in `jj` can be pointers to the same RNG (so, if we're passing it to the CommitBuilder, it's even more important to keep it an `Arc`). Otherwise, I'm worried some part of `jj` will use an RNG different from the others and it will restart from the initial seed. (One silver lining would be that this would have to happen in a test, and thus would hopefully be caught soon, though it may cause someone to waste quite a bit of debugging time).

Reply:
Could we create the RNG eagerly instead? I don't think I understood why we create it lazily. It's probably not expensive, is it?

Reply:
If we create it eagerly, then setting the `debug.randomness-seed` option with the `--config-toml` command-line option will be impossible. We don't need that now, but it would be a weird inconsistency.

We could create it eagerly and replace it inside the `incorporate_toml_config` function, perhaps. That seems better unless there's some subtle problem with it.

Reply:
Maybe it's okay to recreate rng at `incorporate_toml_config()`? No one would generate a change id before processing `--config-toml`.

If we need to create rng lazily, `once_cell::sync::OnceCell` can also be used.

Reply:
I removed the laziness and followed Yuja's suggestion to pass the RNG to CommitBuilder. Thanks, I think it's a definite improvement!

>  That seems better and simpler; I'll sleep on it to see if I see any problems with that.

I remembered the issue, see the comment in the code. I think I found an better way to mostly avoid it.